### ISSUE TITLE
Cannot navigate on `componentDidMount`

### ISSUE THREAD

**Issue by titouancreach** [router5/router5#341]
Hi,

I have troubles to navigate in the `componentDidMount` hook.
I've forked router5-react example, and rewrote some of the inbox component to redirect after the component have mounted

```jsx
class Inbox extends React.Component {

  componentDidMount() {
    this.props.router.navigate('compose')
    
  }

  render() {
    const { route, emails, router } = this.props

    return (
        <div className="inbox">
            <InboxList emails={emails} />
            {route.name === 'inbox.message' ? (
                <Message {...emails[route.params.id]} key={route.params.id} />
            ) : null}
        </div>
    )
  }
}

export default (props) => (
    <RouteNode nodeName="inbox">
        {({ route, router }) => <Inbox route={route} router={router} {...props} />}
    </RouteNode>
)
```

The first time, (when we arrive on this page from the browser), the URL is correclty updated but the page content doesn't change. The `route.name` is wrong whereas `router.getState().name` is good.

Here is the reproduction: https://stackblitz.com/edit/react-router5-new-context-api-fhgkus?file=components/Inbox.js

Thanks!

**Comment by rlebosse** [comment]
I have the same problem.

**Comment by ematipico** [comment]
Is you main component inside the `RouteProvider`?

**Comment by troch** [comment]
I don't think it is related to the lifecycle hook used. I'm going to have a look.

**Comment by troch** [comment]
So I've checked, and it seems related to stackblitz runtime: if you run it in a separate window (by clicking "open in a new window" on the top right), there isn't an issue

**Comment by titouancreach** [comment]
I'm not sure because I had the same error on Chrome, this is why I created the repro link on stackblitz, can we reopen ?

**Comment by troch** [comment]
I had another look. I must had changed something in the stackblitz example which made me thing it wasn't an issue.

The issue is with `Route` and `RouteNode` subscribing to updates in `componentDidMount`. So due to children component `componentDidMount` being called first, navigation in your component `componentDidMount` happens with no listener listening to route changes (hence why the view doesn't update).

I will move it to the constructor, in the mean time if you need that functionality, you can wrap it with `setTimeout`.

**Comment by titouancreach** [comment]
Thanks @troch  !

**Comment by troch** [comment]
6.3.0 has been released, and fix is confirmed (Don't use 6.2.0, it had an issue).
